diff --git a/package/zfs/zfs.mk b/package/zfs/zfs.mk
index 59cdd31..d0ab118 100644
--- a/package/zfs/zfs.mk
+++ b/package/zfs/zfs.mk
@@ -17,12 +17,10 @@ ZFS_AUTORECONF = YES
 
 ZFS_DEPENDENCIES = libaio openssl udev util-linux zlib libcurl linux
 
-# sysvinit installs only a commented-out modules-load.d/ config file
 ZFS_CONF_OPTS = \
 	--with-linux=$(LINUX_DIR) \
 	--with-linux-obj=$(LINUX_DIR) \
-	--disable-rpath \
-	--disable-sysvinit
+	--disable-rpath
 
 ifeq ($(BR2_PACKAGE_LIBTIRPC),y)
 ZFS_DEPENDENCIES += libtirpc
