diff --git a/package/zfs/zfs.mk b/package/zfs/zfs.mk
index b86214e7..d7695149 100644
--- a/package/zfs/zfs.mk
+++ b/package/zfs/zfs.mk
@@ -14,12 +14,10 @@ ZFS_CPE_ID_PRODUCT = openzfs
 
 ZFS_DEPENDENCIES = libaio openssl udev util-linux zlib libcurl linux
 
-# sysvinit installs only a commented-out modules-load.d/ config file
 ZFS_CONF_OPTS = \
 	--with-linux=$(LINUX_DIR) \
 	--with-linux-obj=$(LINUX_DIR) \
-	--disable-rpath \
-	--disable-sysvinit
+	--disable-rpath
 
 ifeq ($(BR2_PACKAGE_LIBTIRPC),y)
 ZFS_DEPENDENCIES += libtirpc
